plugins/inputs/dcos/dcos.go

```diff
@@ -0,0 +1,453 @@
+	url := m.ClusterURL + fmt.Sprintf(AgentContainerMetricsUrl, agentId, containerId)
+	var metric metric
+	err := m.handleJsonRequest(url, &metric)
+	if err != nil {
+		return err
+	}
+	m.processMetric(&metric, acc, Container)
+	return nil
+}
+
+//gatherContainerAppMetrics collects metric for an app in given container, if available
+func (m *Dcos) gatherContainerAppMetrics(agentId string, containerId string, acc telegraf.Accumulator) error {
+	url := m.ClusterURL + fmt.Sprintf(AgentContainerAppMetricsUrl, agentId, containerId)
+	var metric metric
+
+	err := m.handleJsonRequest(url, &metric)
+	if err != nil {
+		return err
+	}
+
+	m.processMetric(&metric, acc, App)
+
+	return nil
+}
+
+//processMetric validated metric and fills accumulator with metric data
+func (m *Dcos) processMetric(metric *metric, acc telegraf.Accumulator, metricType string) {
+
+	measurementData := m.prepareMetric(metric, metricType, acc)
+	//store current timestamp so all measurements for current set of metrics have the same timestamp
+	now := time.Now()
+	for measurementSuffix, tagPoints := range measurementData {
+		for _, points := range tagPoints {
+			tags := make(map[string]string)
+			fields := make(map[string]interface{})
+
+			fillTags(tags, metric.Dimensions)
+
+			tags["scope"] = metricType
+			tags["cluster_url"] = m.ClusterURL
+
+			for _, dp := range points {
+				switch t := dp.Value.(type) {
+				case string: //field values come as real numbers, string is usually "NaN'
+					acc.AddError(fmt.Errorf("Invalid value for field %s: '%s'", dp.Name, t))
+				default:
+					fields[dp.Name] = t
+					fillTags(tags, dp.Tags)
+				}
+			}
+			//fmt.Println(measurementSuffix, fields, tags)
+			acc.AddFields("dcos_"+measurementSuffix, fields, tags, now)
+		}
+	}
+
+}
+
+//fillTags traverses source map and fills tags with non map values
+func fillTags(tags map[string]string, source map[string]interface{}) {
+	for k, v := range source {
+		var s string
+		if v != "" { //remove tags with empty value
+			switch t := v.(type) {
+			case map[string]interface{}:
+				fillTags(tags, t)
+			default:
+				s = interfaceToString(v)
+				tags[k] = s
+			}
+
+		}
+	}
+}
+
+//prepareMetric sorts datapoints according to prefix and optional tag
+func (m *Dcos) prepareMetric(metric *metric, metricType string, acc telegraf.Accumulator) map[string]map[string][]datapoint {
+	//map measurement->(grouping)tag->datapoints
+	measurementData := make(map[string]map[string][]datapoint)
+	for _, d := range metric.Datapoints {
+		if m.preProcessDataPoint(&d, metricType) {
+			continue
+		}
+		var measurementSuffix string
+		if metricType != App {
+			nameSegs := strings.SplitN(d.Name, ".", 2)
+			if len(nameSegs) == 2 {
+				nameSegs[1] = snakeCaser.Replace(nameSegs[1])
+			} else {
+				//metric name could be already divided  by '_'
+				nameSegs = strings.SplitN(d.Name, "_", 2)
+				if len(nameSegs) != 2 {
```

Comment:
Because they are unorganizable. But looking on this after a while, better solution would be to store them in some generic measurement, e.g. dcos_unsorted, dcos_general.